pkg/util/id.go

```diff
@@ -16,31 +16,81 @@ limitations under the License.
 package util
 
 import (
+	"fmt"
 	"io/ioutil"
+	"os"
+	"path"
 	"strings"
+
+	"github.com/google/uuid"
+)
+
+const (
+	maxNodeIDLength = 12
 )
 
-const maxMachineIDLength = 12
+// To ensure a constant node ID between runs, the id will be persisted in the data directory.
+func LoadPersistedNodeID(dataDir string) (string, error) {
+	id, err := loadNodeID(dataDir)
+	if err != nil {
+		return "", fmt.Errorf("failed to load node id. %+v", err)
+	}
 
-func GetMachineID() (string, error) {
-	buf, err := ioutil.ReadFile("/etc/machine-id")
+	err = saveNodeID(dataDir, id)
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to save node id. %+v", err)
+	}
+
+	return id, nil
+}
+
+// Attempt to load the node from the data dir. If not found,
+// generate a new random id
+func loadNodeID(dataDir string) (string, error) {
+	// first read the id from the cache
+	buf, err := ioutil.ReadFile(path.Join(dataDir, "node-id"))
+	if err == nil {
+		// successfully loaded the id from the file
+		id := strings.TrimSpace(string(buf))
```

Comment:
trimming after reading from the file was just trying to help protect against corruption of the file